<a href="https://colab.research.google.com/github/edgarrr19/SimMat2025/blob/main/Tarea_1_3_Programaci%C3%B3n_lineal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables:

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

In [1]:
import scipy.optimize as opt
import numpy as np
import sympy as sym

Optimización:
$$
\min_{x_1-x_6}(0.013x_1+0.008x_2+0.01x_3+0.002x_4+0.005x_5+0.001x_6),
$$

Restricciones:
En $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

* $x_1+x_2+x_3+x_4+x_5+x_6=100$ gramos totales
* $0.10x_1+0.20x_2+0.15x_3+0.04x_5\geq8=-0.10x_1-0.20x_2-0.15x_3-0.04x_5\leq-8$ gramos de proteína
* $0.08x_1+0.10x_2+0.11x_3+0.01x_4+0.01x_5\geq6=-0.08x_1-0.10x_2-0.11x_3-0.01x_4-0.01x_5\leq-6$ gramos de grasa
* $0.001x_1+0.005x_2+0.005x_3+0.10x_4+0.15x_5\leq2$ gramos de fibra
* $0.002x_1+0.005x_2+0.007x_3+0.002x_4+0.008x_5\leq0.4$ gramos de sal


In [4]:
c=np.array([0.013,0.008,0.010,0.002,0.005,0.001])
A=np.array([[-1,-1,-1,-1,-1,-1],
            [-0.10,-0.20,-0.15,0,-0.04,0],
            [-0.08,-0.10,-0.11, -0.01,-0.01,0],
            [0.001,0.005,0.005,0.10,0.15,0],
            [0.002,0.005,0.007,0.002,0.008,0]])
b=np.array([-100,-8,-6,2,0.4])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  4.000e+00  0.000e+00  1.700e+00
                              1.000e-01]
                 marginals: [-1.000e-03 -0.000e+00 -7.000e

**Resultado**

Se necesitan **60 g de carne** y **40 g de gel**:
* 20% proteína = 12 g
* 10% de grasa = 6 g
* 0.5% de fibra = 0.30 g
* 0.5% de sal = 0.30 g
* El gel no aporta nada nutrimentalmente

Con un precio de **$0.52** cada 100 gramos

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente.


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla.

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

Optimización:
$$
\min_{x_1-x_9}(5x_1+2x_2+7x_3+3x_4+6x_5+6x_6+6x_7+1x_8+2x_9),
$$
Restricciones:
* $x_1+x_2+x_3\leq80$ millones de kW se generan
* $x_4+x_5+x_6\leq40$ millones de kW
* $x_7+x_8+x_9\leq60$ millones de kW
* $x_1+x_4+x_7\geq70$ millones de kW se consumen
* $x_2+x_5+x_8\geq40$ millones de kW
* $x_3+x_6+x_9\geq70$ millones de kW

In [5]:
c=np.array([5,2,7,3,6,6,6,1,2])
A=np.array([[1,1,1,0,0,0,0,0,0],
            [0,0,0,1,1,1,0,0,0],
            [0,0,0,0,0,0,1,1,1],
            [-1,0,0,-1,0,0,-1,0,0],
            [0,-1,0,0,-1,0,0,-1,0],
            [0,0,-1,0,0,-1,0,0,-1]])
b=np.array([80,40,60,-70,-40,-70])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

**Resultado**
El precio mínimo es de **540** u.m.:
+ 30 millones de Kw distribuidos de la Planta 1 a Guadalajara
+ 40 millones de Kw distribuidos de la Planta 1 a León
+ 10 millones de Kw distribuidos de la Planta 1 a Morelia
+ 40 millones de Kw distribuidos de la Planta 2 a Guadalajara
+ 60 millones de Kw distribuidos de la Planta 3 a Morelia

### Fábrica de focos

Una fábrica produce focos incandesentes y los vende a 4.5 u.m., además de focos ahorradores que vende a 6 u.m. cada uno. La producción está limitada a 400 focos incandesentes y 300 focos ahorradores al día, además de no poder producir más de 500 focos en total.

Si la fábrica vende toda su producción, determina cuántos focos incandesentes y cuántos focos ahorradores debe producir para obtener los máximos ingresos posibles y cuáles serían éstos.

* $x_1$: Focos incandesentes
* $x_2$: Focos ahorradores


Optimización:
$$
\min_{x_1,x_2}(4.5x_1+6x_2),
$$
Restricciones:
* $x_1+x_2\leq500$
* $x_1\leq400$
* $x_2\leq300$

In [6]:
c=np.array([-4.5,-6])
A=np.array([[1,1],
            [1,0],
            [0,1]])
b=np.array([500,400,300])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -2700.0
              x: [ 2.000e+02  3.000e+02]
            nit: 1
          lower:  residual: [ 2.000e+02  3.000e+02]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 0.000e+00  2.000e+02  0.000e+00]
                 marginals: [-4.500e+00 -0.000e+00 -1.500e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado**

La ganancia máxima será de **$2700.00** produciendo:
* 200 focos incandesentes
* 300 focos ahorradores

### Problema de compra al mayoreo
Un frutero necesita 16 cajas de naranjas, 5 de plátanos y 20 de manzanas. Dos mayoristas pueden suministrarle para satisfacer sus necesidades, pero sólo venden la fruta en contenedores completos. El mayorista A envía en cada contenedor 8 cajas de naranjas, 1 de plátanos y 2 de manzanas. El mayorista B envía en cada contenedor 2 cajas de naranjas, una de plátanos y 7 de manzanas. Sabiendo que cada contenedor del mayorista A tiene un costo de 150 u.m. y el contenedor del mayorista B de 300 u.m., calcular cuántos contenedores habrá de comprar a cada mayorista, con objeto de reducir al mínimo el costo de lo solicitado.

* $x_1$: naranjas A (cajas)
* $x_2$: platanos A
* $x_3$: manzanas A
* $x_4$: naranjas B (cajas)
* $x_5$: platanos B
* $x_6$: manzanas B

El frutero necesita $16x_{1,4}+5x_{2,5}+20x_{3,6}$

Optimización:
$$
\min_{x_A,x_B}(150x_A+300x_B),
$$
Restricciones:
* $8x_1+2x_4\geq16$ naranjas
* $x_2+x_5\geq5$ platanos
* $2x_3+7x_6\geq20$ manzanas




In [12]:
c = np.array([150, 300])
A = np.array([[-8, -2],   # naranjas (A,B)
              [-1, -1],   # plátanos
              [-2, -7]])  # manzanas
b = np.array([-16, -5, -20])
opt.linprog(c,A,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 1050.0
              x: [ 3.000e+00  2.000e+00]
            nit: 3
          lower:  residual: [ 3.000e+00  2.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 1.200e+01  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -9.000e+01 -3.000e+01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado**

Para cumplir la demanda mínima se necesitan:
* **3** contenedores del proovedor **A**
* **2** contenedores del proovedor **B**
* Sobran **12** naranjas

Teniendo un costo total de $1050 u.m.